File: python/04_read_from_shared_file.py
import threading

def read_from_shared_file(file):
    with open(file, "r") as f:
        print("Thread {} read from file: {}".format(threading.current_thread().name, f.read()))

if __name__ == '__main__':
    file = "shared_file.txt"

    for i in range(10):
        t = threading.Thread(target=read_from_shared_file, args=(file,))
        t.start()

File: C#/04_read_from_shared_file.cs
using System;
using System.IO;
using System.Threading;

class Program {
    static void Main() {
        string filePath = "shared_file.txt";
        var fileContent = File.ReadAllText(filePath);

        var threads = new Thread[10];
        for (int i = 0; i < threads.Length; i++) {
            threads[i] = new Thread(() => ReadFile(filePath));
            threads[i].Start();
        }

        foreach (var thread in threads) {
            thread.Join();
        }
    }

    static void ReadFile(string filePath) {
        var threadId = Thread.CurrentThread.ManagedThreadId;
        var content = File.ReadAllText(filePath);
        Console.WriteLine("Thread {0} read from file: {1}", threadId, content);
    }
}
